Use event name to break timestamp ties

JustinKuli opened the pull request:
It is possible for two events to be emitted in the same millisecond. If this happens, the last timestamps can't be used to consistently order the events. However, events emitted by client-go's recorder are suffixed with a hexadecimal nanosecond timestamp, which might be used to order these events.

Refs:
 - https://github.com/stolostron/backlog/issues/26039

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>

JustinKuli committed:
Use event name to break timestamp ties

It is possible for two events to be emitted in the same millisecond. If
this happens, the last timestamps can't be used to consistently order
the events. However, events emitted by client-go's recorder are suffixed
with a hexadecimal nanosecond timestamp, which might be used to order
these events.

Refs:
 - https://github.com/stolostron/backlog/issues/26039

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>

JustinKuli committed:
Add event timestamp collision tests

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>

ChunxiAlexLuo commented on controllers/sync/policy_status_sync.go:
```diff
@@ -258,6 +259,32 @@ func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Requ
 		}
 		// sort by lasttimestamp
 		sort.Slice(history, func(i, j int) bool {
+			if history[i].LastTimestamp.String() == history[j].LastTimestamp.String() {
```
Based on our discussion, this fix looks good to me and holds for the other reviewer.

1. How about one line info log here for the successful cases that indicate we made the additional comparison?
2. Refer https://github.com/stolostron/backlog/issues/26066  if that issue is the future punch list?

ChunxiAlexLuo reviewed (dismissed):
/lgtm
/hold

JustinKuli committed:
Use microsecond timestamp (EventTime) if present

Newer versions of events will have this more precise timestamp that can
be used to order the events when the "normal" timestamps collide.

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>

JustinKuli commented:
@mprahl iirc you wondered if we could store the more precise timestamp in the policy status, but since it's stored as a `metav1.Time`, I think that would require a corresponding change in the policy type definition in the propagator repo.

mprahl reviewed (dismissed):
Nice job!

JustinKuli commented:
Added the log suggested by @ChunxiAlexLuo. If you're happy with the changes, I think we can approve and un-hold this.

JustinKuli commented:
/unhold

ChunxiAlexLuo reviewed (approved):
/lgtm

JustinKuli commented:
I feel like I've seen this before (but it's rare)... retesting

```
• [FAILED] [31.881 seconds]
Test status sync [AfterEach]
/home/runner/work/governance-policy-status-sync/governance-policy-status-sync/governance-policy-status-sync/test/e2e/case2_status_sync_test.go:51
  Should set status to compliant
  /home/runner/work/governance-policy-status-sync/governance-policy-status-sync/governance-policy-status-sync/test/e2e/case2_status_sync_test.go:70

  Begin Captured GinkgoWriter Output >>
    STEP: Creating a policy on hub cluster in ns:other-namespace 10/04/22 20:14:55.911
    STEP: Creating a policy on managed cluster in ns:managed 10/04/22 20:14:56.396
    STEP: Generating an compliant event on the policy 10/04/22 20:14:56.616
    STEP: Checking if policy status is compliant 10/04/22 20:14:56.621
    STEP: Checking if hub policy status is in sync 10/04/22 20:14:57.628
    STEP: Deleting a policy on hub cluster in ns:other-namespace 10/04/22 20:14:57.631
  << End Captured GinkgoWriter Output

  Timed out after 30.001s.
  Expected
      <*errors.errorString | 0xc000103e30>: {
          s: "list size doesn't match, expected 0 actual 1",
      }
  to be nil
```
